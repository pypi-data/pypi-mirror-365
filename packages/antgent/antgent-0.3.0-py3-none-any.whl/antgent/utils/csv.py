import csv
import io
from typing import Any, cast

from openpyxl import Workbook


def _get_nesting_cols_by_name(nesting_keys: list[str], fieldnames: list[str]) -> list[str]:
    """Get nesting columns by name."""
    actual_nesting_cols = []
    fieldnames_lower = {f.lower(): f for f in fieldnames}
    for key in nesting_keys:
        actual_col = fieldnames_lower.get(str(key).lower())
        if not actual_col:
            raise ValueError(f"CSV must contain column '{key}'.")
        actual_nesting_cols.append(actual_col)
    return actual_nesting_cols


def _get_nesting_cols_by_index(nesting_keys: list[int], fieldnames: list[str]) -> list[str]:
    """Get nesting columns by index."""
    actual_nesting_cols = []
    for index in nesting_keys:
        try:
            actual_nesting_cols.append(fieldnames[int(index)])
        except IndexError as e:
            raise IndexError(f"Column index {index} is out of range.") from e
    return actual_nesting_cols


def _get_actual_nesting_cols(nesting_keys: list[str] | list[int], fieldnames: list[str]) -> list[str]:
    """Resolves nesting keys to actual column names from fieldnames."""
    if not nesting_keys:
        return []

    if isinstance(nesting_keys[0], str):
        return _get_nesting_cols_by_name(cast(list[str], nesting_keys), fieldnames)
    if isinstance(nesting_keys[0], int):
        return _get_nesting_cols_by_index(cast(list[int], nesting_keys), fieldnames)

    raise TypeError("nesting_keys must be a list of strings or a list of integers.")


def _build_nested_structure(reader: csv.DictReader, actual_nesting_cols: list[str]) -> dict:
    """Builds the nested dictionary from the CSV reader."""
    result = {}
    for row in reader:
        current_level = result
        # Iterate through nesting keys to build the nested structure
        for i, col_name in enumerate(actual_nesting_cols):
            key_value = row.pop(col_name)

            if i < len(actual_nesting_cols) - 1:
                # This is not the last nesting level, so ensure a dict exists
                if key_value not in current_level:
                    current_level[key_value] = {}
                current_level = current_level[key_value]
            else:
                # This is the last nesting level, assign the rest of the row
                current_level[key_value] = row
    return result


def csv_to_nested_dict(csv_string: str, nesting_keys: list[str] | list[int]) -> dict:
    """
    Converts a CSV string to a nested dictionary based on a specified list of nesting keys.

    The nesting keys determine the structure of the output dictionary.
    Keys can be column names (str) or column indices (int).
    Column name matching for nesting keys is case-insensitive.
    """
    f = io.StringIO(csv_string.strip())
    reader = csv.DictReader(f)

    if not reader.fieldnames:
        return {}

    actual_nesting_cols = _get_actual_nesting_cols(nesting_keys, list(reader.fieldnames))
    if not actual_nesting_cols:
        return {}

    return _build_nested_structure(reader, actual_nesting_cols)


def list_dict_to_csv(data: list[dict]) -> str:
    """
    Converts a list of dictionaries to a CSV formatted string.

    The keys of the first dictionary in the list are used as the header row.
    Handles values with commas and other special characters by quoting them.

    Args:
        data: A list of dictionaries. Each dictionary represents a row.
              It's assumed that all dictionaries have the same keys.

    Returns:
        A string containing the data in CSV format.
        Returns an empty string if the input list is empty.
    """
    if not data:
        return ""

    # Use an in-memory string buffer to build the CSV string.
    # The `newline=''` argument is important to prevent extra blank rows.
    output = io.StringIO()

    # The fieldnames are the keys from the first dictionary.
    fieldnames = data[0].keys()

    # Create a DictWriter object. This object can map dictionaries to CSV rows.
    # The `quoting=csv.QUOTE_MINIMAL` (the default) ensures that fields are only
    # quoted if they contain the delimiter, quotechar, or any of the characters
    # in lineterminator.
    writer = csv.DictWriter(output, fieldnames=fieldnames)

    # Write the header row (the dictionary keys).
    writer.writeheader()

    # Write all the dictionary data to the CSV buffer.
    writer.writerows(data)

    # Return the complete CSV string from the buffer.
    return output.getvalue()


def list_dict_to_xlsx_bytes(data: list[dict[str, Any]]) -> bytes:
    """
    Converts a list of dictionaries to an XLSX file as bytes.

    Args:
        data: A list of dictionaries. Each dictionary represents a row.
              It's assumed that all dictionaries have the same keys.

    Returns:
        The XLSX file content as bytes.
        Returns empty bytes if the input list is empty.
    """
    if not data:
        return b""

    workbook = Workbook()
    sheet = workbook.active

    # Write headers
    headers = list(data[0].keys())
    for col_num, header_title in enumerate(headers, 1):
        cell = sheet.cell(row=1, column=col_num)
        cell.value = header_title

    # Write data rows
    for row_num, row_data in enumerate(data, 2):
        for col_num, header in enumerate(headers, 1):
            cell = sheet.cell(row=row_num, column=col_num)
            cell.value = row_data.get(header)

    # Save to a bytes buffer
    buffer = io.BytesIO()
    workbook.save(buffer)
    buffer.seek(0)
    return buffer.getvalue()
