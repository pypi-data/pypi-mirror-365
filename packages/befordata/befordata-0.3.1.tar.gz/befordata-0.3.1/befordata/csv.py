import gzip as _gzip
import lzma as _lzma
from io import StringIO as _StringIO
from pathlib import Path as _Path
from typing import List as _List

import pandas as _pd

from ._record import BeForRecord


def read_csv(
    file_path: str | _Path,
    columns: str | _List[str] | None = None,
    encoding: str = "utf-8",
    comment_char: str = "#",
):
    """Reads CSV file

    The function can handle comments as well as compressed CSVs, if they end
    with `.csv.xz` or `.csv.gz`

    Parameters
    ----------
    file_path : str | Path
        the path to the CSV file. If file end with `.csv.xz` or `.csv.gz`,
        decompression will be used
    columns : str | list[str], Optional
        the columns that should be read. If no columns are specified all columns
        are read
    encoding : str, optional
        file encoding, default="utf-8",

    comment_char : str, Optional
        line starting with character or string will be treated as comments and
        returned as a list of strings.

    Returns
    -------
    pandas.DataFrame and list[str] with all comments
    """

    p = _Path(file_path)
    if p.suffix.endswith("xz"):
        fl = _lzma.open(p, "rt", encoding=encoding)
    elif p.suffix.endswith("gz"):
        fl = _gzip.open(file_path, "rt", encoding=encoding)
    else:
        fl = open(file_path, "r", encoding=encoding)

    csv_str = ""
    comments = []
    for l in fl.readlines():
        if l.startswith(comment_char):
            comments.append(l)
        else:
            csv_str += l
    fl.close()

    df = _pd.read_csv(_StringIO(csv_str))
    if isinstance(columns, str):
        columns = [columns]
    if isinstance(columns, list):
        df = df.loc[:, columns]

    return df, comments


def read_csv_as_befordata(
    file_path: str | _Path,
    sampling_rate: float,
    columns: str | _List[str] | None = None,
    sessions: _List[int] | None = None,
    time_column: str | None = None,
    meta: dict | None = None,
    encoding: str = "utf-8",
    comment_char: str = "#",
):
    """Read CSV file as befordata

    The function can handle comments as well as compressed CSVs, if they end
    with `.csv.xz` or `.csv.gz`

    Parameters
    ----------
    file_path: _description_
    sampling_rate: _description_
    columns: _description_. Defaults to None.
    sessions: _description_. Defaults to None.
    time_column: _description_. Defaults to None.
    meta: _description_. Defaults to None.
    encoding: _description_. Defaults to "utf-8".
    comment_char: _description_. Defaults to "#".

    Returns
    -------
    BeForRecord

    """

    df, _ = read_csv(file_path=file_path, encoding=encoding, comment_char=comment_char)
    if columns is None:
        columns = []
    elif isinstance(columns, str):
        columns = [columns]

    if time_column is None:
        time_column = ""
    if sessions is None:
        sessions = []
    if not isinstance(meta, dict):
        meta = {}

    return BeForRecord(
        dat=df,
        sampling_rate=sampling_rate,
        sessions=sessions,
        time_column=time_column,
        meta=meta,
    )
