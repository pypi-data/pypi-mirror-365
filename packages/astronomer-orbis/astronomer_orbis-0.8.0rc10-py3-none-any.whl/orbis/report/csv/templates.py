"""CSV template management and basic row operations."""

import csv
import logging
import os
from typing import Any

logger = logging.getLogger("root")


def get_csv_template_headers() -> list[str]:
    """Get the headers from the CSV template file."""
    template_path = os.path.join(os.path.dirname(__file__), "..", "..", "config", "csv_template.csv")
    try:
        with open(template_path) as f:
            reader = csv.reader(f)
            return next(reader)  # Get the header row
    except Exception as e:
        logger.error(f"Failed to read CSV template: {e}")
        raise


def get_empty_row() -> dict[str, Any]:
    """Get an empty row with just the headers."""
    return dict.fromkeys(get_csv_template_headers(), "")
