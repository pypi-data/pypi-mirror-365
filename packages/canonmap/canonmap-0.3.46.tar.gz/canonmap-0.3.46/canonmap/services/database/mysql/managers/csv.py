# canonmap/services/database/mysql/managers/csv.py
import logging
logger = logging.getLogger(__name__)

import csv
from pathlib import Path
from typing import Union, List, Dict, Any, Tuple
import pandas as pd

from canonmap.services.database.mysql.config import MySQLConfig
from canonmap.services.database.mysql.utils.create_mysql_ddl import create_mysql_ddl

class CSVManager:
    def __init__(self, connection_manager: MySQLConfig):
        self.connection_manager = connection_manager

    def create_table_ddl_from_csv(
        self,
        csv_path_or_buffer,
        table_name: str,
        save_dir: Union[None, str, Path] = None
    ) -> str:
        return create_mysql_ddl(csv_path_or_buffer, table_name, save_dir)

    def _smart_fix_csv(
        self,
        csv_path: Union[str, Path],
        fixed_csv_path: Union[str, Path] = "fixed.csv",
        bad_rows_path: Union[str, Path] = "bad_rows.csv",
        log_bad_rows: bool = True
    ) -> Tuple[str, List[Tuple[int, str]]]:
        """
        Reads a CSV, identifies and separates malformed rows, and writes a clean version.
        A row is considered malformed if it has a different number of columns than the header.
        It attempts to fix rows with more columns by merging adjacent fields.
        
        Returns:
            Tuple[str, List[Tuple[int, str]]]: (fixed_csv_path, bad_rows_data)
        """
        csv_path = Path(csv_path)
        fixed_csv_path = Path(fixed_csv_path)
        bad_rows_path = Path(bad_rows_path)
        
        bad_rows_data = []

        try:
            with open(csv_path, 'r', newline='', encoding='utf-8') as infile:
                # Read header to determine expected number of columns
                try:
                    header = next(csv.reader(infile))
                    expected_cols = len(header)
                except StopIteration:
                    logger.warning(f"CSV file '{csv_path}' is empty.")
                    # Create empty fixed file and return
                    fixed_csv_path.touch()
                    return str(fixed_csv_path), []

                # Rewind and process whole file
                infile.seek(0)
                reader = csv.reader(infile)
                next(reader) # Skip header

                with open(fixed_csv_path, 'w', newline='', encoding='utf-8') as outfile:
                    writer = csv.writer(outfile)
                    writer.writerow(header)

                    for lineno, row in enumerate(reader, start=2):
                        if len(row) == expected_cols:
                            writer.writerow(row)
                        elif len(row) > expected_cols:
                            # Heuristic: try to merge fields to fix it.
                            # This merges the second field with subsequent fields to meet the expected column count.
                            num_extra_cols = len(row) - expected_cols
                            merged_row = row[:1] + [','.join(row[1:1 + num_extra_cols + 1])] + row[1 + num_extra_cols + 1:]
                            if len(merged_row) == expected_cols:
                                writer.writerow(merged_row)
                                logger.debug(f"Fixed line {lineno} by merging columns.")
                            else:
                                # This case should ideally not be reached with this logic
                                bad_rows_data.append((lineno, ','.join(row)))
                        else: # len(row) < expected_cols
                            # This is a malformed row, log it. 
                            # Don't pad, as it can hide data integrity issues.
                            bad_rows_data.append((lineno, ','.join(row)))
            
            if bad_rows_data and log_bad_rows:
                logger.warning(f"Found {len(bad_rows_data)} malformed row(s). Writing them to '{bad_rows_path}'.")
                with open(bad_rows_path, 'w', newline='', encoding='utf-8') as bad_file:
                    writer = csv.writer(bad_file)
                    writer.writerow(['line_number', 'original_content'])
                    for lineno, content in bad_rows_data:
                        writer.writerow([lineno, content])

            logger.info(f"✅ Finished cleaning. Clean CSV written to: {fixed_csv_path}")
            return str(fixed_csv_path), bad_rows_data

        except Exception as e:
            logger.error(f"Error processing CSV file '{csv_path}': {e}")
            raise

    def process_csv_to_json(
        self,
        csv_path: Union[str, Path],
        table_name: str
    ) -> Tuple[List[Dict[str, Any]], str]:
        """
        Process a CSV file: clean it, convert to JSON format, and generate DDL statement.
        
        Args:
            csv_path (Union[str, Path]): Path to the CSV file
            table_name (str): Name of the table for DDL generation
            
        Returns:
            Tuple[List[Dict[str, Any]], str]: (json_data, ddl_statement)
        """
        try:
            # Use a more descriptive name for the bad rows file.
            bad_rows_csv_path = Path(csv_path).stem + "_bad_rows.csv"
            fixed_csv_path, bad_rows = self._smart_fix_csv(
                csv_path, 
                bad_rows_path=bad_rows_csv_path,
                log_bad_rows=True
            )
            
            if not Path(fixed_csv_path).exists() or Path(fixed_csv_path).stat().st_size == 0:
                logger.warning(f"CSV file '{csv_path}' is empty or could not be processed. Nothing to load.")
                return [], ""

            # Generate DDL statement
            ddl_statement = self.create_table_ddl_from_csv(fixed_csv_path, table_name)
            
            # Convert CSV to JSON format
            import pandas.errors
            logger.info(f"Converting CSV '{fixed_csv_path}' to JSON format...")
            try:
                # Read the CSV, keeping blank values as empty strings and handling NaNs
                df = pd.read_csv(fixed_csv_path, keep_default_na=False, na_values=[''])

                # Replace any numpy NaN values with None for SQL compatibility (NULL)
                df = df.astype(object).where(pd.notnull(df), None)
                
                # Convert DataFrame to list of dictionaries
                json_data = df.to_dict('records')
                
                logger.info(f"✅ Converted {len(json_data)} rows to JSON format")
                return json_data, ddl_statement

            except pandas.errors.ParserError as e:
                logger.error(f"ParserError while reading cleaned CSV '{fixed_csv_path}': {e}")
                raise
            except Exception as e:
                logger.error(f"An unexpected error occurred while reading the CSV: {e}")
                raise
            
        except Exception as e:
            logger.error(f"Error while processing CSV '{csv_path}': {e}")
            raise
