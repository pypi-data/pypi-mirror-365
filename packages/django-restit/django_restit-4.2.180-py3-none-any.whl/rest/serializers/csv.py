from objict import objict
from io import StringIO
from django.http import StreamingHttpResponse
import csv
from rest import helpers as rh
from . import localizers


def flattenObject(obj, field_names, localize=None):
    row = []
    for f in field_names:
        d = ""
        if "__" in f:
            f = f.replace("__", ".")
        if "." in f:
            # we could use obj.getFieldValue
            keys = f.split('.')
            f1 = keys[0]
            f2 = keys[1]
            if f1 == "metadata" and hasattr(obj, "getProperty"):
                if len(keys) > 2:
                    d = obj.getProperty(keys[2], category=f2)
                else:
                    d = obj.getProperty(f2)
            elif hasattr(obj, "getFieldValue"):
                d = obj.getFieldValue(f, "t/a")
                if callable(d):
                    d = d()
            else:
                d1 = getattr(obj, f1, None)
                if d1 is not None:
                    if not hasattr(d1, f2):
                        if hasattr(d1, "first"):
                            d1 = d1.first()
                    d = getattr(d1, f2, "")
                    if callable(d):
                        d = d()               
        else:
            d = getattr(obj, f, "n/a")
            if callable(d):
                d = d()
        if d is None:
            d = "n/a"
        elif hasattr(d, "pk"):
            d = d.pk
        if localize and f in localize:
            rh.debug("localize....", localize[f])
            d = localizeValue(d, localize[f])
        else:
            d = str(d)
        row.append(d)
    return row


def localizeValue(value, localizer_name, extra=None):
    if "|" in localizer_name:
        localizer_name, extra = localizer_name.split("|")
    localizer = getattr(localizers, localizer_name, None)
    if localizer is not None:
        return localizer(value, extra)
    return value


def extractFieldNames(fields):
    header = []
    field_names = []
    for f in fields:
        if type(f) is tuple:
            r, f = f
            field_names.append(r)
        else:
            field_names.append(f)
        header.append(f)
    return header, field_names


def generateCSV(qset, fields, name, header_cols=None,
                values_list=False, output=None, stream=False, localize=None):
    a = objict()
    a.name = name
    a.file = StringIO()
    if output:
        a.file = output
    csvwriter = csv.writer(a.file)
    header, field_names = extractFieldNames(fields)
    if header_cols:
        header = header_cols
    csvwriter.writerow(header)
    if values_list:
        for row in qset.values_list(*field_names):
            row = [str(x) for x in row]
            csvwriter.writerow(row)
    else:
        rh.debug("localize.", localize)
        for obj in qset:
            csvwriter.writerow(flattenObject(obj, field_names, localize))
    if hasattr(a.file, "getvalue"):
        a.data = a.file.getvalue()
    a.mimetype = "text/csv"
    return a


def iterCsvObject(items, writer, header, field_names, localize=None):
    yield writer.writerow(header)
    for item in items:
        yield writer.writerow(flattenObject(item, field_names, localize))


def generateCSVStream(qset, fields, name, localize=None):
    import csv
    # check if we support stream mode
    header, field_names = extractFieldNames(fields)
    # rows = qset.values_list(*fields)
    pseudo_buffer = EchoWriter()
    writer = csv.writer(pseudo_buffer)
    return StreamingHttpResponse(
        iterCsvObject(qset, writer, header, field_names, localize=localize))


class EchoWriter(object):
    """An object that implements just the write method of the file-like
    interface.
    """
    def writeline(self, value):
        return "{}\n".format(value)

    def write(self, value):
        """Write the value by returning it, instead of storing in a buffer."""
        return value
