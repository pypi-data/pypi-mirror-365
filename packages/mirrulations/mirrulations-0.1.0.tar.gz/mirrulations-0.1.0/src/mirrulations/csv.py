import os
import json
import csv
import click
from tqdm import tqdm


def resolve_output_path(output_option, docket_id):
    import os
    if output_option:
        if os.path.isdir(output_option) or output_option.endswith(os.sep):
            return os.path.join(output_option, f"{docket_id}.csv")
        else:
            return output_option
    else:
        return os.path.join(os.getcwd(), f"{docket_id}.csv")

@click.command("csv")
@click.argument('input_dir', type=click.Path(exists=True, file_okay=False, readable=True))
@click.option('-o', '--output', type=click.Path(), default=None, help='Output CSV file path or directory')
@click.option('--dryrun', is_flag=True, help='Only report which fields will be included or excluded, do not write CSV')
@click.option('-include', multiple=True, help='Force-include a field that would otherwise be omitted due to constant value')
def main(input_dir, output, dryrun, include):
    """
    Convert a directory of JSON files (all with the same structure) to a CSV file.
    """
    # Extract docketID from the first file
    first_file = next((f for f in os.listdir(input_dir) if f.endswith('.json')), None)
    if not first_file:
        click.echo('No JSON files found in input directory.')
        return
    with open(os.path.join(input_dir, first_file), 'r', encoding='utf-8') as f:
        try:
            data = json.load(f)
        except Exception as e:
            click.echo(f'Error reading {first_file}: {e}')
            return
        # Verify this is a comment file
        if 'data' not in data or 'attributes' not in data['data'] or 'type' not in data['data'] or data['data']['type'] != 'comments':
            click.echo('Error: Directory does not contain comment data.')
            return
        docket_id = data['data']['attributes']['docketId']
    output_path = resolve_output_path(output, docket_id)
    click.echo(f'Output CSV will be: {output_path}')
    # Step 1: Scan all JSON files to determine field stats
    attribute_fields = set()
    null_fields = set()
    constant_fields = dict()  # field -> value
    variable_fields = set()
    field_values = dict()  # field -> set of seen values
    total_files = 0
    input_files = [f for f in os.listdir(input_dir) if f.endswith('.json')]
    if not input_files:
        click.echo('No JSON files found in input directory.')
        return
    for filename in tqdm(input_files, desc='Scanning JSON files'):
        total_files += 1
        with open(os.path.join(input_dir, filename), 'r', encoding='utf-8') as f:
            try:
                data = json.load(f)
            except Exception as e:
                click.echo(f'Error reading {filename}: {e}')
                continue
            attributes = data['data'].get('attributes', {})
            for field, value in attributes.items():
                attribute_fields.add(field)
                if field not in field_values:
                    field_values[field] = set()
                field_values[field].add(json.dumps(value, sort_keys=True))
    # Analyze fields
    for field in attribute_fields:
        values = field_values[field]
        if all(json.loads(v) is None for v in values):
            null_fields.add(field)
        elif len(values) == 1:
            constant_fields[field] = json.loads(next(iter(values)))
        else:
            variable_fields.add(field)
    # Apply -include override
    include_set = set(include)
    for field in include_set:
        if field in constant_fields:
            variable_fields.add(field)
            constant_fields.pop(field)
    # Step 2: Report field analysis
    click.echo('\n--- Field Analysis Report ---')
    click.echo('Fields to be included:')
    for field in sorted(["id", "api_url"] + list(variable_fields)):
        click.echo(f'  {field}')
    click.echo('Fields excluded because all values were null:')
    for field in sorted(null_fields):
        click.echo(f'  {field}')
    click.echo('Fields excluded because all values were the same:')
    for field, value in constant_fields.items():
        click.echo(f'  {field}: {repr(value)}')

    if dryrun:
        return
    # Step 3: Write CSV
    csv_fields = ["id", "api_url"] + sorted(variable_fields)
    with open(output_path, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.DictWriter(csvfile, fieldnames=csv_fields)
        writer.writeheader()
        for filename in tqdm(input_files, desc='Writing CSV'):
            with open(os.path.join(input_dir, filename), 'r', encoding='utf-8') as f:
                try:
                    data = json.load(f)
                except Exception as e:
                    click.echo(f'Error reading {filename}: {e}')
                    continue
                row = {}
                row['id'] = data['data'].get('id', '')
                row['api_url'] = data['data'].get('links', {}).get('self', '')
                attributes = data['data'].get('attributes', {})
                for field in variable_fields:
                    row[field] = attributes.get(field, '')
                writer.writerow(row)

if __name__ == "__main__":
    main()
