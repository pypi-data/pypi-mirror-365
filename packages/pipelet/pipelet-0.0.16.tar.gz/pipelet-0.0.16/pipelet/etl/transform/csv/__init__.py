from pipelet.etl.transform.csv.reader import CsvParser

__all__ = ("CsvParser",)
