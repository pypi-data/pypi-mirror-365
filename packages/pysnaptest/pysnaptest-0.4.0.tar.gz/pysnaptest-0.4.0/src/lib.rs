#![deny(clippy::unwrap_used)]

use pyo3::{
    pymodule,
    types::{PyModule, PyModuleMethods},
    wrap_pyfunction, Bound, PyResult,
};

mod common;
mod errors;
mod mocks;

pub use common::*;
pub use errors::*;
pub use mocks::*;

use std::{collections::HashMap, path::PathBuf};

use csv::ReaderBuilder;
use pyo3::prelude::*;

#[pyfunction]
#[pyo3(signature = (test_info, result, redactions=None))]
pub fn assert_json_snapshot(
    test_info: &SnapshotInfo,
    result: &Bound<'_, PyAny>,
    redactions: Option<HashMap<String, RedactionType>>,
) -> PyResult<()> {
    let res: serde_json::Value = pythonize::depythonize(result)?;
    let snapshot_name = test_info.snapshot_name();
    let mut settings: insta::Settings = test_info.try_into()?;

    for (selector, redaction) in redactions.unwrap_or_default() {
        settings.add_redaction(selector.as_str(), redaction);
    }

    settings.bind(|| {
        insta::assert_json_snapshot!(snapshot_name, res);
    });
    Ok(())
}

#[pyfunction]
#[pyo3(signature = (test_info, result, redactions=None))]
pub fn assert_csv_snapshot(
    test_info: &SnapshotInfo,
    result: &str,
    redactions: Option<HashMap<String, RedactionType>>,
) -> PyResult<()> {
    let mut rdr = ReaderBuilder::new().from_reader(result.as_bytes());
    let columns: Vec<Vec<serde_json::Value>> = vec![rdr
        .headers()
        .expect("Expects csv with headers")
        .into_iter()
        .map(|h| h.into())
        .collect()];
    let records = rdr
        .into_deserialize()
        .collect::<Result<Vec<Vec<serde_json::Value>>, _>>()
        .expect("Failed to parse csv records");
    let res: Vec<Vec<serde_json::Value>> = columns.into_iter().chain(records).collect();

    let snapshot_name = test_info.snapshot_name();
    let mut settings: insta::Settings = test_info.try_into()?;

    for (selector, redaction) in redactions.unwrap_or_default() {
        settings.add_redaction(selector.as_str(), redaction);
    }

    settings.bind(|| {
        insta::assert_csv_snapshot!(snapshot_name, res);
    });
    Ok(())
}

#[pyfunction]
pub fn assert_binary_snapshot(
    test_info: &SnapshotInfo,
    extension: &str,
    result: Vec<u8>,
) -> PyResult<()> {
    let snapshot_name = test_info.snapshot_name();
    let settings: insta::Settings = test_info.try_into()?;
    settings.bind(|| {
        insta::assert_binary_snapshot!(format!("{snapshot_name}.{extension}").as_str(), result);
    });
    Ok(())
}

#[pyfunction]
pub fn assert_snapshot(test_info: &SnapshotInfo, result: &Bound<'_, PyAny>) -> PyResult<()> {
    let snapshot_name = test_info.snapshot_name();
    let settings: insta::Settings = test_info.try_into()?;
    settings.bind(|| {
        insta::assert_snapshot!(snapshot_name, result);
    });
    Ok(())
}

#[pymethods]
impl SnapshotInfo {
    #[staticmethod]
    #[pyo3(signature = (snapshot_path_override = None, snapshot_name_override = None, allow_duplicates = false))]
    fn from_pytest(
        snapshot_path_override: Option<PathBuf>,
        snapshot_name_override: Option<String>,
        allow_duplicates: bool,
    ) -> PyResult<Self> {
        Ok(
            if let (Some(snapshot_folder), Some(snapshot_name)) = (
                snapshot_path_override.clone(),
                snapshot_name_override.clone(),
            ) {
                Self {
                    snapshot_folder,
                    snapshot_name,
                    relative_test_file_path: None,
                    allow_duplicates,
                }
            } else {
                let pytest_info: SnapshotInfo = PytestInfo::from_env()?.try_into()?;
                Self {
                    snapshot_folder: snapshot_path_override.unwrap_or(pytest_info.snapshot_folder),
                    snapshot_name: snapshot_name_override.map_or(pytest_info.snapshot_name, |v| {
                        v.split('-').next().map_or(v.clone(), |s| s.to_string())
                    }),
                    relative_test_file_path: pytest_info.relative_test_file_path,
                    allow_duplicates,
                }
            },
        )
    }

    pub fn snapshot_folder(&self) -> &PathBuf {
        &self.snapshot_folder
    }

    pub fn last_snapshot_name(&self) -> String {
        let test_idx = Self::counters()
            .get(&self.snapshot_name)
            .cloned()
            .unwrap_or(1);
        self.snapshot_name_with_idx(test_idx)
    }

    pub fn next_snapshot_name(&self) -> String {
        let test_idx = Self::counters()
            .get(&self.snapshot_name)
            .cloned()
            .unwrap_or(0)
            + 1;
        self.snapshot_name_with_idx(test_idx)
    }

    pub fn last_snapshot_path(&self, module_path: Option<String>) -> PyResult<PathBuf> {
        let module_path = module_path
            .unwrap_or(module_path!().to_string())
            .replace("::", "__");
        Ok(self.snapshot_folder.join(format!(
            "{module_path}__{}@pysnap.snap",
            self.last_snapshot_name()
        )))
    }

    pub fn next_snapshot_path(&self, module_path: Option<String>) -> PyResult<PathBuf> {
        let module_path = module_path
            .unwrap_or(module_path!().to_string())
            .replace("::", "__");
        Ok(self.snapshot_folder.join(format!(
            "{module_path}__{}@pysnap.snap",
            self.next_snapshot_name()
        )))
    }
}

#[pymodule]
#[pyo3(name = "_pysnaptest")]
fn pysnaptest(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_class::<SnapshotInfo>()?;

    m.add_function(wrap_pyfunction!(assert_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_binary_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_json_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(assert_csv_snapshot, m)?)?;
    m.add_function(wrap_pyfunction!(mock_json_snapshot, m)?)?;
    m.add_class::<PySnapshot>()?;
    Ok(())
}
