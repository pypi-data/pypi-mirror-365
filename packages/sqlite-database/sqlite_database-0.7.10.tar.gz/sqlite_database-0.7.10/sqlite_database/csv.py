"""CSV module, used to export database/table to csv"""

from csv import DictWriter, DictReader
from io import StringIO
from os import mkdir
from os.path import join as join_path, isfile, exists

from . import Database, Table


def _export_table(table: Table) -> str:
    data = StringIO()
    tb_data = table.select_one()
    if tb_data is None:
        return ""
    fields = tuple(tb_data.keys())
    csv_writer = DictWriter(data, fields)
    csv_writer.writeheader()
    for selecteds in table.paginate_select(length=1):
        if len(selecteds) == 0:
            break
        selected = selecteds[0]
        csv_writer.writerow(selected)
    return data.getvalue()


def _export_database(database: Database) -> list[tuple[str, str]]:
    exported = []
    for table in database.tables():
        exported.append((table.name, _export_table(table)))
    return exported


def _process_database(filename: str, data: list[tuple[str, str]]):
    for table in data:
        base = join_path(filename, f"{table[0]}.csv")
        with open(base, "w", encoding="utf-8") as file:
            file.write(table[1])


def to_csv_string(table_or_database: Table | Database):
    """Export database/table to csv"""
    if isinstance(table_or_database, Table):
        return _export_table(table_or_database)
    if isinstance(table_or_database, Database):
        return _export_database(table_or_database)
    raise TypeError(
        f"Expected Table or Database, got {type(table_or_database).__name__}"
    )


def to_csv_file(table_or_database: Table | Database, file: str):
    """Export database/table to csv file, will act different if database is exported.

    If database is passed, it tries to create a directory then export all table"""
    readed = to_csv_string(table_or_database)
    if isinstance(readed, list):
        if not exists(file):
            mkdir(file)
        if isfile(file):
            raise NotADirectoryError(f"{file} is a file, export failed!")

        _process_database(file, readed)
        return True
    with open(file, "w", encoding="utf-8") as fio:
        fio.write(readed)
    return True

def from_csv_string(table: Table, data: str):
    """Insert from CSV data"""
    reader = DictReader(data)
    for entry in reader:
        table.insert(entry)

def from_csv_file(table: Table, data: str):
    """Insert from CSV file"""
    with open(data, encoding='utf-8') as f:
        fdata = f.read()
    return from_csv_string(table, fdata)
