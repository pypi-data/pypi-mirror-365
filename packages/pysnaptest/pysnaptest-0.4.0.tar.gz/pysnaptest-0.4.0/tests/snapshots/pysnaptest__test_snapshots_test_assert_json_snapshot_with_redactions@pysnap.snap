---
source: src/lib.rs
description: "Test File Path: tests/test_snapshots.py"
---
{
  "also_level_one": "[redacted]",
  "level_one": "left_alone"
}
